### ISSUE TITLE
Support for Private Repositories?

### ISSUE THREAD

**Issue by shivams** [rdehuyss/micropython-ota-updater#2]
Are you planning on supporting private repos? Or is it already implemented?

**Comment by milkshakespoon** [comment]
I'd also like to see this feature if possible! I'm guessing this would require some auth for GitHub, which I don't see in the current code...

**Comment by sergiuszm** [comment]
ota_updater.py
```
class OTAUpdater:

    def __init__(self, github_repo, module='', main_dir='src', headers={}):
        self.http_client = HttpClient(headers=headers)
        self.github_repo = github_repo.rstrip('/').replace('https://github.com', 'https://api.github.com/repos')
        self.main_dir = main_dir
        self.module = module.rstrip('/')


class HttpClient:

    def __init__(self, headers={}):
        self._headers = headers

    def request(self, method, url, data=None, json=None, headers={}, stream=None):
        def _write_headers(sock, _headers):
            for k in _headers:
                sock.write(b'{}: {}\r\n'.format(k, _headers[k]))

        try:
            proto, dummy, host, path = url.split('/', 3)
        except ValueError:
            proto, dummy, host = url.split('/', 2)
            path = ''
        if proto == 'http:':
            port = 80
        elif proto == 'https:':
            import ussl
            port = 443
        else:
            raise ValueError('Unsupported protocol: ' + proto)

        if ':' in host:
            host, port = host.split(':', 1)
            port = int(port)

        ai = usocket.getaddrinfo(host, port, 0, usocket.SOCK_STREAM)
        ai = ai[0]

        s = usocket.socket(ai[0], ai[1], ai[2])
        try:
            s.connect(ai[-1])
            if proto == 'https:':
                s = ussl.wrap_socket(s, server_hostname=host)
            s.write(b'%s /%s HTTP/1.0\r\n' % (method, path))
            if not 'Host' in headers:
                s.write(b'Host: %s\r\n' % host)
            # Iterate over keys to avoid tuple alloc
            _write_headers(s, self._headers)
            _write_headers(s, headers)

            # add user agent
            s.write('User-Agent')
            s.write(b': ')
            s.write('MicroPython OTAUpdater')
            s.write(b'\r\n')
            if json is not None:
                assert data is None
                import ujson
                data = ujson.dumps(json)
                s.write(b'Content-Type: application/json\r\n')
            if data:
                s.write(b'Content-Length: %d\r\n' % len(data))
            s.write(b'\r\n')
            if data:
                s.write(data)

            l = s.readline()
            # print(l)
            l = l.split(None, 2)
            status = int(l[1])
            reason = ''
            if len(l) > 2:
                reason = l[2].rstrip()
            while True:
                l = s.readline()
                if not l or l == b'\r\n':
                    break
                # print(l)
                if l.startswith(b'Transfer-Encoding:'):
                    if b'chunked' in l:
                        raise ValueError('Unsupported ' + l)
                elif l.startswith(b'Location:') and not 200 <= status <= 299:
                    raise NotImplementedError('Redirects not yet supported')
        except OSError:
            s.close()
            raise

        resp = Response(s)
        resp.status_code = status
        resp.reason = reason
        return resp

    def head(self, url, **kw):
        return self.request('HEAD', url, **kw)

    def get(self, url, **kw):
        return self.request('GET', url, **kw)

    def post(self, url, **kw):
        return self.request('POST', url, **kw)

    def put(self, url, **kw):
        return self.request('PUT', url, **kw)

    def patch(self, url, **kw):
        return self.request('PATCH', url, **kw)

    def delete(self, url, **kw):
        return self.request('DELETE', url, **kw)
```

main.py
```
    token='YOUR_GITHUB_TOKEN'
    updater = OTAUpdater('https://github.com/sergiuszm/cae_fipy', headers={'Authorization': 'token {}'.format(token)})
```

**Comment by reichley** [comment]
> ota_updater.py
> 
> ```
> class OTAUpdater:
> 
>     def __init__(self, github_repo, module='', main_dir='src', headers={}):
>         self.http_client = HttpClient(headers=headers)
>         self.github_repo = github_repo.rstrip('/').replace('https://github.com', 'https://api.github.com/repos')
>         self.main_dir = main_dir
>         self.module = module.rstrip('/')
> 
> 
> class HttpClient:
> 
>     def __init__(self, headers={}):
>         self._headers = headers
> 
>     def request(self, method, url, data=None, json=None, headers={}, stream=None):
>         def _write_headers(sock, _headers):
>             for k in _headers:
>                 sock.write(b'{}: {}\r\n'.format(k, _headers[k]))
> 
>         try:
>             proto, dummy, host, path = url.split('/', 3)
>         except ValueError:
>             proto, dummy, host = url.split('/', 2)
>             path = ''
>         if proto == 'http:':
>             port = 80
>         elif proto == 'https:':
>             import ussl
>             port = 443
>         else:
>             raise ValueError('Unsupported protocol: ' + proto)
> 
>         if ':' in host:
>             host, port = host.split(':', 1)
>             port = int(port)
> 
>         ai = usocket.getaddrinfo(host, port, 0, usocket.SOCK_STREAM)
>         ai = ai[0]
> 
>         s = usocket.socket(ai[0], ai[1], ai[2])
>         try:
>             s.connect(ai[-1])
>             if proto == 'https:':
>                 s = ussl.wrap_socket(s, server_hostname=host)
>             s.write(b'%s /%s HTTP/1.0\r\n' % (method, path))
>             if not 'Host' in headers:
>                 s.write(b'Host: %s\r\n' % host)
>             # Iterate over keys to avoid tuple alloc
>             _write_headers(s, self._headers)
>             _write_headers(s, headers)
> 
>             # add user agent
>             s.write('User-Agent')
>             s.write(b': ')
>             s.write('MicroPython OTAUpdater')
>             s.write(b'\r\n')
>             if json is not None:
>                 assert data is None
>                 import ujson
>                 data = ujson.dumps(json)
>                 s.write(b'Content-Type: application/json\r\n')
>             if data:
>                 s.write(b'Content-Length: %d\r\n' % len(data))
>             s.write(b'\r\n')
>             if data:
>                 s.write(data)
> 
>             l = s.readline()
>             # print(l)
>             l = l.split(None, 2)
>             status = int(l[1])
>             reason = ''
>             if len(l) > 2:
>                 reason = l[2].rstrip()
>             while True:
>                 l = s.readline()
>                 if not l or l == b'\r\n':
>                     break
>                 # print(l)
>                 if l.startswith(b'Transfer-Encoding:'):
>                     if b'chunked' in l:
>                         raise ValueError('Unsupported ' + l)
>                 elif l.startswith(b'Location:') and not 200 <= status <= 299:
>                     raise NotImplementedError('Redirects not yet supported')
>         except OSError:
>             s.close()
>             raise
> 
>         resp = Response(s)
>         resp.status_code = status
>         resp.reason = reason
>         return resp
> 
>     def head(self, url, **kw):
>         return self.request('HEAD', url, **kw)
> 
>     def get(self, url, **kw):
>         return self.request('GET', url, **kw)
> 
>     def post(self, url, **kw):
>         return self.request('POST', url, **kw)
> 
>     def put(self, url, **kw):
>         return self.request('PUT', url, **kw)
> 
>     def patch(self, url, **kw):
>         return self.request('PATCH', url, **kw)
> 
>     def delete(self, url, **kw):
>         return self.request('DELETE', url, **kw)
> ```
> 
> main.py
> 
> ```
>     token='YOUR_GITHUB_TOKEN'
>     updater = OTAUpdater('https://github.com/sergiuszm/cae_fipy', headers={'Authorization': 'token {}'.format(token)})
> ```

Thanks for the code update to be able to use this with private repositories!

**Comment by ebolisa** [comment]
How about if the repo is on my own domain like "https://www.mydomain.com/updates" ?

**Comment by bachi76** [comment]
If @rdehuyss accepts PRs the code from @sergiuszm would make a nice one..

**Comment by rdehuyss** [comment]
Definitely open for PR's though I cannot test myself.

Can anybody verify the solution of @sergiuszm? If so, I'll update it.

**Comment by sergiuszm** [comment]
> 
> 
> Definitely open for PR's though I cannot test myself.
> 
> Can anybody verify the solution of @sergiuszm? If so, I'll update it.

I have used the code for my own deployed in the field units without any issues for over a year.

**Comment by rdehuyss** [comment]
Resolved with commit 074f1c620709fa6eb53e1392605cee88a60def94
